jina/serve/stream/__init__.py

```diff
@@ -206,15 +219,43 @@ def iterating_task_done(task):
 
         iterate_requests_task.add_done_callback(iterating_task_done)
 
-        while not all_requests_handled.is_set():
-            future = await result_queue.get()
-            try:
-                response = self._result_handler(future.result())
+        async def receive_responses():
+            while not all_requests_handled.is_set():
+                future = await result_queue.get()
+                try:
+                    response = self._result_handler(future.result())
+                    yield response
+                    requests_to_handle.count -= 1
+                    update_all_handled()
+                except self._EndOfStreaming:
+                    pass
+
+        async for response in receive_responses():
+            if not results_in_order:
+                # if is not needed to return_in_order, just yield responses as they come.
                 yield response
-                requests_to_handle.count -= 1
-                update_all_handled()
-            except self._EndOfStreaming:
-                pass
+            else:
+                # if return_in_order is needed, we have to check if the response is to be returned
+                response_request_id = await DataRequest.extract_id(response)
+                if response_request_id == request_ids[0]:
+                    # is the response's turn to be yielded
+                    yield response
+                    request_ids.pop(0)
+                else:
+                    # is not your turn, put to the end
+                    responses_ids[response_request_id] = len(responses_list)
+                    responses_list.append(response)
+
+                stop_yielding = False
+                while not stop_yielding and len(request_ids) > 0:
+                    next_request_id = request_ids[0]
+                    if next_request_id in responses_ids:
+                        response_index = responses_ids[next_request_id]
+                        del responses_ids[next_request_id]
+                        request_ids.pop(0)
+                        yield responses_list[response_index]
```

Comment:
assuming we no more need responses_list, we can change the code like so:
```suggestion
                        request_ids.pop(0)
                        yield responses_ids.pop(next_request_id)
```